### PR TITLE
Updating the APEX package

### PR DESCRIPTION
**Pull request by khuck**
Removed old, not-supported version 0.1 and added latest official
release of APEX.  Added all dependencies.  Added maintainer.

### CODE REVIEW

File: var/spack/repos/builtin/packages/apex/package.py [hunk 1]
```diff
@@ -4,30 +4,78 @@
...
+    variant('papi', default=darwin_default, description='Enables PAPI support')
+    variant('gperftools', default=False, description='Enables Google PerfTools TCMalloc support')
+    variant('jemalloc', default=False, description='Enables JEMalloc support')
+
+    # Disable by default
+    variant('cuda', default=False, description='Enables CUDA support')
+    variant('boost', default=False, description='Enables Boost support')
+
+    # Dependencies
+    depends_on('cmake')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/apex/package.py]:
```suggestion
    depends_on('cmake', type='build')
```

File: var/spack/repos/builtin/packages/apex/package.py [hunk 2]
```diff
@@ -4,30 +4,78 @@
...
-    depends_on("activeharmony@4.5:")
-    depends_on("ompt-openmp")
+    # Disable some default dependencies on Darwin/OSX
+    darwin_default = False
+    if sys.platform != 'darwin':
+        darwin_default = True
+
+    # Enable by default
+    variant('activeharmony', default=True, description='Enables Active Harmony support')
+    variant('binutils', default=True, description='Enables Binutils support')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/apex/package.py:34]: What kind of binutils support is this? All packages use binutils, but we try not to add a dependency on binutils to every package.

File: var/spack/repos/builtin/packages/apex/package.py [hunk 3]
```diff
@@ -4,30 +4,78 @@
...
+            args.append('-DUSE_OMPT=TRUE')
+        if '+gperftools' in spec:
+            args.append('-DGPERFTOOLS_ROOT={0}'.format(
+                spec['gperftools'].prefix))
+            args.append('-DUSE_TCMALLOC=TRUE')
+        if '+jemalloc' in spec:
+            args.append('-DJEMALLOC_ROOT={0}'.format(spec['jemalloc'].prefix))
+            args.append('-DUSE_JEMALLOC=TRUE')
+        if '+boost' in spec:
+            args.append('-DBOOST_ROOT={0}'.format(spec['boost'].prefix))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/apex/package.py]: I would explicitly disable these in an else-clause whenever possible.

File: var/spack/repos/builtin/packages/apex/package.py (hunk 2 above)
Comment (khuck) [reply var/spack/repos/builtin/packages/apex/package.py:34]: APEX uses bfd to resolve function pointer addresses, and demangle.h for demangling.  Actually, this should be similar to the TAU dependency, so I just now realize I need to add some variants to that dependency...

File: var/spack/repos/builtin/packages/apex/package.py (hunk 3 above)
Comment (khuck) [reply var/spack/repos/builtin/packages/apex/package.py]: will do.  thanks for the review!

File: var/spack/repos/builtin/packages/apex/package.py (hunk 2 above)
Comment (khuck) [reply var/spack/repos/builtin/packages/apex/package.py:34]: oops, just realized I do have the depends_on correct already, disregard the comment about additional changes to that dependency...

File: var/spack/repos/builtin/packages/apex/package.py [hunk 4]
```diff
@@ -4,30 +4,77 @@
...
 
 class Apex(CMakePackage):
     """Autonomic Performance Environment for eXascale (APEX)."""
 
-    homepage = "http://github.com/khuck/xpress-apex"
-    url      = "http://github.com/khuck/xpress-apex/archive/v0.1.tar.gz"
+    maintainers = ['khuck']
+    homepage = "https://github.com/khuck/xpress-apex"
+    url      = "https://github.com/khuck/xpress-apex/archive/v2.2.0.tar.gz"
+    git      = "https://github.com/khuck/xpress-apex"
```
Comment (darmac) [inline var/spack/repos/builtin/packages/apex/package.py:16]: I think we don't need `git` here. `url` is enough for this package.

**Comment by darmac** [comment]
@khuck  I can install it succesfully now(by default variant)

**Comment by DavidPoliakoff** [comment]
Any advice from the Spack team on what would need to happen for this to merge?

File: var/spack/repos/builtin/packages/apex/package.py [hunk 5]
```diff
@@ -4,30 +4,141 @@
...
+        else:
+            args.append('-D' + prefix + '_JEMALLOC=FALSE')
+
+        if '+boost' in spec:
+            args.append('-DBOOST_ROOT={0}'.format(spec['boost'].prefix))
+
+        if '+tests' in spec:
+            args.append('-D' + test_prefix + 'BUILD_TESTS=TRUE')
+        else:
+            args.append('-D' + test_prefix + 'BUILD_TESTS=FALSE')
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/apex/package.py]: Here you can do:
```suggestion
        args.append(self.define_from_variant('APEX_BUILD_TESTS', 'tests'))
```
and similarly for all the other ones. If it makes sense you can construct a list literal that way:
```python
args = [
    self.define_from_variant(...),
    self.define_from_variant(...),
    self.define_from_variant(...)
]
```

**Review by alalazo — CHANGES_REQUESTED** [review]
Left some comments, let me know what you think about it.

File: var/spack/repos/builtin/packages/apex/package.py (hunk 5 above)
Comment (khuck) [reply var/spack/repos/builtin/packages/apex/package.py]: done for all cmake variables, thanks

**Comment by khuck** [comment]
@alalazo all changes addressed, I think

File: lib/spack/docs/packaging_guide.rst [hunk 6]
```diff
@@ -3147,7 +3147,8 @@ as arguments.
 
 Here are the definitions of the three built-in flag handlers:
 
-.. code-block:: python
+   def build_system_flags(self, name, flags):
+       return (None, None, flags)
```
Comment (alalazo) [inline lib/spack/docs/packaging_guide.rst]: Is this a change that got in inadvertently?

**Review by alalazo — CHANGES_REQUESTED** [review]
Just a minor comment

File: lib/spack/docs/packaging_guide.rst (hunk 6 above)
Comment (khuck) [reply lib/spack/docs/packaging_guide.rst]: I have no idea where that came from.  If you like, I can re-clone the main repo and change just apex/package.py and push that, without all the merges.

Comment (khuck) [reply lib/spack/docs/packaging_guide.rst]: I pushed a clean commit, removing this weird merge history from the last 8 months.

**Review by alalazo — APPROVED** [review]

**Pull request merged**